lib/nerves_hub/client.ex

```diff
@@ -0,0 +1,47 @@
+defmodule NervesHub.Client do
+  @moduledoc """
+  Responsible for determining if an update should be applied.
+  """
+  require Logger
+
+  @typedoc "Update that comes over a socket."
+  @type update_data :: map()
+
+  @callback update_available(update_data) :: :apply | :ignore | {:reschedule, pos_integer()}
+
+  @callback reboot_required() :: :apply | :ignore | {:reschedule, pos_integer()}
```

Comment:
Is there a case where reboots aren't required after a firmware update? Making them optional makes me uneasy since it opens up the possibility of trying to update multiple times w/o rebooting and that won't work.

Reply:
I'd have to verify, but the `fwup.conf`s for all of the official systems fails if a reboot is pending. Early in the project, I didn't do this and it was easier to accidentally double update. If a mandatory reboot is always done at the end of every update, then you don't have to worry about the case. There are scenarios where you might want to delay a reboot, but I think we can handle those in a separate pass on this - especially when there's a use case to warrant it.

Reply:
Also `fwup` can fail an update for a lot of reasons. I think that @mobileoverlord was wanting to add in handling for those failures in a separate pass so that it didn't complicate the "delay an update" case.

Reply:
I am okay with the idea that you can delay an update, but once you commit, the system is going to reboot.